descriptors/web_app.js

```diff
@@ -0,0 +1,174 @@
+    for (characteristic of characteristics) {
+      switch (characteristic.uuid) {
+        case testCharacteristicUUID:
+          let descriptors = await characteristic.getDescriptors();
+          assertEquals(1, descriptors.length);
+          const descriptor = await characteristic.getDescriptor(
+            BluetoothUUID.getDescriptor('gatt.characteristic_user_description'));
+          assertEquals(
+            BluetoothUUID.getDescriptor('gatt.characteristic_user_description'),
+            descriptor.uuid);
+          assertEquals(characteristic, descriptor.characteristic,
+            'descriptor characteristic mismatch');
+          assertEquals(null, descriptor.value, 'value should be null');
+          const value = await descriptor.readValue();
+          const decoder = new TextDecoder('utf-8');
+          let strVal = decoder.decode(value);
+          assertEquals('Test UINT32 characteristic', strVal);
+
+          strVal = decoder.decode(value);
+          assertEquals('Test UINT32 characteristic', strVal);
+
+          const encoder = new TextEncoder('utf-8')
+          try {
+            await descriptor.writeValue(encoder.encode('Something else'));
+          } catch (e) {
+            assertEquals('NotSupportedError', e.name);
+          }
+          break;
+        default:
+          logError(`Unexpected characteristic: ${characteristic.uuid}`);
+      }
+    }
+  }
+
+  // Verify descriptors for Nordic UART Service (NUS)
+  // https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/include/bluetooth/services/nus.html
+  const verifyNordicServiceDescriptors = async (service) => {
+    assertEquals(nordicUARTService, service.uuid, 'incorrect service UUID');
+
+    const characteristics = await service.getCharacteristics();
+    assertEquals(2, characteristics.length, 'expected 2 characteristic');
+    for (characteristic of characteristics) {
+      switch (characteristic.uuid) {
+        case '6e400002-b5a3-f393-e0a9-e50e24dcca9e': // RX
+          try {
+            await characteristic.getDescriptors();
+            throw 'RX Characteristic should have no descriptors';
+          } catch (e) {
+            assertEquals('NotFoundError', e.name);
+          }
+          break;
+        case '6e400003-b5a3-f393-e0a9-e50e24dcca9e': // TX
+          const descriptors = await characteristic.getDescriptors();
+          assertEquals(1, descriptors.length);
+          const descriptor = descriptors[0];
+          assertEquals(
+            BluetoothUUID.getDescriptor('gatt.client_characteristic_configuration'),
+            descriptor.uuid);
+          assertEquals(characteristic, descriptor.characteristic,
+            'descriptor characteristic mismatch');
+          const value = await descriptor.readValue();
+          const decoder = new TextDecoder('utf-8');
+          const strVal = decoder.decode(value);
+          assertEquals(2, strVal.length);
+          assertEquals('\0\0', strVal);
```

Comment:
This is a weird value for the descriptor to have. I'm not sure we should bake it into our tests.